### PR TITLE
Create reusable component for list of actions

### PR DESCRIPTION
**Pull request by TW80000**
Required for https://nanopay.atlassian.net/browse/CPF-966.

## Changes

* Moved the logic that generates the context menu in the table view into it's own view so that it can be reused.
  * In particular, I want to use it in the `tableCellFormatter` for properties on models being displayed in the table. For example, on an Invoice model that has a list of attachments, we can generate a dropdown action list view where each action opens the attachment in a new tab.
* Updated the "three vertical dots" icon to match the designs.

### Resting and disabled are the same (as specified by design)
Screen Shot 2019-04-17 at 11 44 35 AM

### Hover (disabled entries won't transition to this state)
Screen Shot 2019-04-17 at 11 44 48 AM

### Active
Screen Shot 2019-04-17 at 11 44 55 AM

### The new view being reused in a different tableCellFormatter
Screen Shot 2019-04-17 at 11 46 23 AM

### CODE REVIEW

File: src/foam/u2/view/OverlayActionListView.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    'foam.u2.md.OverlayDropdown'
+  ],
+
+  imports: [
+    'ctrl'
+  ],
+
+  properties: [
+    {
+      type: 'Action[]',
```
Comment (adamvy) [inline src/foam/u2/view/OverlayActionListView.js]: We still have to specify class: 'FObjectArray', class: 'FObjectProperty', class: 'Boolean', etc for properties at the moment.

In the future all you will have to specify is type: '', but that's not fully hooked up yet.

File: src/foam/u2/view/OverlayActionListView.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+      var overlay = this.OverlayDropdown.create();
+      this.active_$.follow(overlay.opened$);
+      this.disabled_$.follow(foam.core.ExpressionSlot.create({
+        args: this.data.map((a) => a.createIsAvailable$(this.obj$)),
+        code: (...rest) => ! rest.reduce((l, r) => l || r, false)
+      }));
+
+      this.
+        callIf(this.data.length > 0, function() {
```
Comment (adamvy) [inline src/foam/u2/view/OverlayActionListView.js]: No need for `callIf(length > 0` just to protect a `forEach(this.data`  If there's no items in data, forEach won't iterate anything.

File: src/foam/u2/view/OverlayActionListView.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+          addClass('noselect').
+          start().
+            addClass(this.myClass('icon')).
+            style({ 'background-image': this.imageURL_$ }).
+            on('mouseover', this.onMouseOver).
+            on('mouseout', this.onMouseOut).
+            callIf(this.data.length > 0, function() {
+              this.on('click', function(evt) {
+                if ( self.disabled_ ) return;
+                overlay.open(evt.clientX, evt.clientY);
```
Comment (adamvy) [inline src/foam/u2/view/OverlayActionListView.js]: I would move building the overlay contents down here.  No sense building all the action views for the overlay until we're about to render it.  Otherwise we're spending all the work every row in the table, but the user is unlikely to click on any particular row.

File: src/foam/u2/view/OverlayActionListView.js [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+          // container with `overflow: hidden` then this overlay won't be cut
+          // off.
+          this.ctrl.add(overlay);
+        }).
+
+        start('span').
+          addClass('noselect').
+          start().
+            addClass(this.myClass('icon')).
+            style({ 'background-image': this.imageURL_$ }).
```
Comment (adamvy) [inline src/foam/u2/view/OverlayActionListView.js:138]: If we're not going to be dynamic on data, we probably don't need to be dynamic on imageURL_ either.

File: src/foam/u2/view/OverlayActionListView.js [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+      cursor: pointer;
+    }
+  `,
+
+  methods: [
+    function initE() {
+      var self = this;
+
+      var overlay = this.OverlayDropdown.create();
+      this.active_$.follow(overlay.opened$);
```
Comment (adamvy) [inline src/foam/u2/view/OverlayActionListView.js]: Need to make sure we .onDetach() these bindings, otherwise we'll be leaking memory.

File: src/foam/u2/view/OverlayActionListView.js (hunk 4 above)
Comment (adamvy) [reply src/foam/u2/view/OverlayActionListView.js:138]: Oh NVM, I see that imageURL_ is in fact dynamic.

File: src/foam/u2/view/OverlayActionListView.js [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        }).
+
+        start('span').
+          addClass('noselect').
+          start().
+            addClass(this.myClass('icon')).
+            style({ 'background-image': this.imageURL_$ }).
+            on('mouseover', this.onMouseOver).
+            on('mouseout', this.onMouseOut).
+            callIf(this.data.length > 0, function() {
```
Comment (adamvy) [inline src/foam/u2/view/OverlayActionListView.js:139]: If we have no actions to display, we should consider not hooking up any listeners including mouseOut/mouseOver and make the button look disabled.

Then we barely have to do any work for rows with no actions.

File: src/foam/u2/view/TableView.js [hunk 7]
```diff
@@ -70,13 +70,13 @@ foam.CLASS({
...
 
-    ^vertDots {
+    ^ .vertDots {
       font-size: 20px;
       font-weight: bold;
       padding-right: 12px;
     }
 
-    ^noselect {
+    ^ .noselect {
```
Comment (adamvy) [inline src/foam/u2/view/TableView.js]: Why the change?

CSS engines are typically faster at matching explicit class names like `^noselect`, rather than having to match on parent/child class name relationships like `^ .noselect`

File: src/foam/u2/view/UnstyledTableView.js [hunk 8]
```diff
@@ -206,11 +199,11 @@ foam.CLASS({
...
                     }).
-                    add(' ', view.vertMenuIcon).
-                    addClass(view.myClass('vertDots')).
-                    addClass(view.myClass('noselect'));
+                    tag(view.Image, { data: '/images/Icon_More_Resting.svg' }).
+                    addClass('vertDots').
+                    addClass('noselect');
                   }).
-                  style({ width: 40 }).
+                  style({ width: 60 }).
```
Comment (adamvy) [inline src/foam/u2/view/UnstyledTableView.js]: I think this should actually be width: '60px'

File: src/foam/u2/view/UnstyledTableView.js [hunk 9]
```diff
@@ -276,47 +274,14 @@ foam.CLASS({
...
-                          overlay.open(evt.clientX, evt.clientY);
-                        });
-                      }).
-                      add(view.vertMenuIcon).
-                    end().
-                  end();
-                });
+                start('td').
+                  addClass(view.myClass('context-menu-cell')).
+                  tag({
```
Comment (adamvy) [inline src/foam/u2/view/UnstyledTableView.js]: The style I would like to move to for `tag(` and `start(` calls like this is to add
```
requires: [
  'foam.u2.view.OverlayActionListView'
]
```

and do

```
.tag(this.OverlayActionListView, { data: actions, obj: obj })
```

This will make the build tool's job easier.

File: src/foam/u2/view/OverlayActionListView.js (hunk 1 above)
Comment (TW80000) [reply src/foam/u2/view/OverlayActionListView.js]: I'll change it, but it seems to work anyway the way it is. Any idea why that would be?

Comment (adamvy) [reply src/foam/u2/view/OverlayActionListView.js]: Because you essentially just had a regular Property axiom installed, instead of a more type specific Array property.  The base property still works as a property, it just doesn't do any of the extra functionality like type checking and adapting that the more specific property types do.  And if you're not relying on that actual functionality then it will all still work.

You can even specify properties as just a name:

```
properties: [
  'data'
]
```

This creates a plain property named data, which still works like a property.  We use things like
```
properties: [
  {
    class: 'Array'
    name: 'prop'
  }
]
```
Because Array property adds things like a useful default factory and type checking if you try to set it to a non-property.

Comment (TW80000) [reply src/foam/u2/view/OverlayActionListView.js]: Got it, thank you!

File: src/foam/u2/view/TableView.js (hunk 7 above)
Comment (TW80000) [reply src/foam/u2/view/TableView.js]: I made that change because I want to use that CSS class in both `UnstyledTableView` and the `OverlayActionListView` but `^noselect` will tie it to the table view.

Maybe this should be moved to our global styles.

File: src/foam/u2/view/OverlayActionListView.js [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+      name: 'disabledImageURL',
+      value: 'images/Icon_More_Disabled.svg'
+    },
+    {
+      type: 'URL',
+      name: 'imageURL_',
+      expression: function(restingImageURL, hoverImageURL, disabledImageURL, activeImageURL, hovered_, disabled_, active_) {
+        if ( disabled_ ) {
+          return `url(${disabledImageURL})`;
+        } else if ( active_ ) {
```
Comment (kgrgreer) [inline src/foam/u2/view/OverlayActionListView.js]: No need for the else's if you've already returned in the if.

File: src/foam/u2/view/TableView.js (hunk 7 above)
Comment (adamvy) [reply src/foam/u2/view/TableView.js]: I see.

The downside of sharing CSS like this is that OverlayActionListView is now no longer fully componentized, it _has_ to be rendered inside a TableView in order for it to behave properly, and this dependency is not obvious in any way to a reader of either model.

Personally I would rather just repeat the CSS in OverlayActionListView, _or_ add some top level FOAM stylesheet that is _always_ included and contains some utility classes like .noselect

I don't think we have such a global stylesheet yet, but it might be worth investigating if we have some CSS classes that we want to reuse like this.

File: src/foam/u2/view/OverlayActionListView.js [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+          end().
+        end();
+    },
+
+    function initializeOverlay() {
+      var self = this;
+      this.overlay_.forEach(this.data, function(action) {
+        this.
+          start().
+            show(action.createIsAvailable$(self.ConstantSlot.create({
```
Comment (adamvy) [inline src/foam/u2/view/OverlayActionListView.js]: `self.obj$` instead of `self.ConstantSlot.create...` would be shorter to type, but it would also create a dynamic binding on self.  If self.obj isn't ever going to change then I guess this works :).

File: src/foam/u2/view/OverlayActionListView.js [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+              value: self.obj
+            }))).
+            addClass(self.myClass('action')).
+            add(action.label).
+            on('click', function(evt) {
+              self.overlay_.close();
+              action.maybeCall(self.obj.__subContext__, self.obj);
+            }).
+            attrs({
+              disabled: action.createIsEnabled$(self.obj$).map(function(e) {
```
Comment (adamvy) [inline src/foam/u2/view/OverlayActionListView.js:166]: Wait why are we using self.obj$ here but a ConstantSlot above?  We should be consistent.  `self.obj$` is probably the most idiomatic and straightforward to read.

Comment (TW80000) [reply src/foam/u2/view/OverlayActionListView.js:166]: Updated to make them consistent.

File: src/foam/u2/view/TableView.js (hunk 7 above)
Comment (TW80000) [reply src/foam/u2/view/TableView.js]: Good point about dependencies between components. I just copied the CSS to the new view and reverted the relevant changes to CSS selectors for the TableView.

**Comment by adamvy** [comment]
Looks great!
